src/traversals/dfs.jl

```diff
@@ -202,3 +264,82 @@ function dfs_tree(g::AbstractGraph, s::Integer)
     end
     return h
 end
+
+# """
+# dfs_tree(g, s)
+
+# Return an ordered vector of vertices representing a directed acylic graph based on
+# depth-first traversal of the graph `g` starting with source vertex `s`.
+# """
+# dfs_tree(g::AbstractGraph, s::Integer; dir=:out) = tree(dfs_parents(g, s; dir=dir))
+
+"""
+dfs_parents(g, s[; dir=:out])
+
+Perform a depth-first search of graph `g` starting from vertex `s`.
+Return a vector of parent vertices indexed by vertex. If `dir` is specified,
+use the corresponding edge direction (`:in` and `:out` are acceptable values).
+
+### Implementation Notes
+This version of DFS is iterative.
+"""
+dfs_parents(g::AbstractGraph, s::Integer; dir=:out) =
+(dir == :out) ? _dfs_parents(g, s, out_neighbors) : _dfs_parents(g, s, in_neighbors)
+
+function _dfs_parents(g::AbstractGraph, s::Integer, neighborfn::Function)
+    T = eltype(g)
+    parents = zeros(T, nv(g))
+
+    seen = falses(nv(g))
+    S = Vector{T}([s])
+    seen[s] = true
+    parents[s] = s
+    while !isempty(S)
+        v = S[end]
+        u = 0
+        for n in neighborfn(g, v)
+            if !seen[n]
+                u = n
+                break
+            end
+        end
+        if u == 0
+            pop!(S)
+        else
+            seen[u] = true
+            push!(S, u)
+            parents[u] = v
+        end
+    end
+    return parents
+end
+dfs_tree2(g::AbstractGraph, s::Integer; dir=:out) = tree(dfs_parents(g, s; dir=dir))
+
+function _dfs_parents3(g::AbstractGraph, s::Integer, neighborfn::Function)
+    T = eltype(g)
+    parents = zeros(T, nv(g))
+
+    seen = zeros(Bool, nv(g))
+    S = Vector{T}([s])
+    seen[s] = true
+    parents[s] = s
+    while !isempty(S)
+        v = S[end]
+        u = 0
+        for n in neighborfn(g, v)
+            if !seen[n]
+                u = n
+                break
+            end
+        end
+        if u == 0
+            pop!(S)
+        else
+            seen[u] = true
+            push!(S, u)
+            parents[u] = v
+        end
+    end
+    return parents
+end
```

Comment:
L293 of _dfs_parents creates a bitvector; L322 of _dfs_parents3 creates a vector of Bool. The latter is faster.